verification/src/compiler/fetcher.rs

```diff
@@ -1,13 +1,22 @@
-use super::version::CompilerVersion;
+use super::version::Version;
 use async_trait::async_trait;
 use std::path::PathBuf;
+use thiserror::Error;
 
-#[async_trait]
-pub trait Fetcher {
-    type Error: Send + Sync + 'static;
-    async fn fetch(&self, ver: &CompilerVersion) -> Result<PathBuf, Self::Error>;
+#[derive(Error, Debug)]
+pub enum FetchError {
+    #[error("version {0} not found")]
+    NotFound(Version),
+    #[error("couldn't fetch the file: {0}")]
+    Fetch(anyhow::Error),
+    #[error("couldn't create file: {0}")]
+    File(std::io::Error),
+    #[error("tokio sheduling error: {0}")]
+    Shedule(tokio::task::JoinError),
 }
 
-pub trait VersionList {
-    fn all_versions(&self) -> Vec<CompilerVersion>;
+#[async_trait]
+pub trait Fetcher: Send + Sync + 'static {
+    async fn fetch(&self, ver: &Version) -> Result<PathBuf, FetchError>;
+    fn all_versions(&self) -> Vec<Version>;
 }
```

Comment:
What is the reason behind removing Error as associated type?

Reply:
Because then we'll still use `dyn Fetcher<Error=FetchErrror>`, and if we're using that, why not just include it into the trait